python/cudf/cudf/core/series.py

```diff
@@ -1652,19 +1652,42 @@ def label_encoding(self, cats, dtype=None, na_sentinel=-1):
         if dtype is None:
             dtype = min_scalar_type(len(cats), 8)
 
-        cats = Series(cats).astype(self.dtype)
+        cats = Series(cats)
+        try:
+            # Where there is a type-cast from string to numeric types,
+            # there is a possibility for ValueError when strings
+            # are having non-numeric values, in such cases we have
+            # to catch the exception and return a encoded labels
+            # with na_sentinel values as there would be no corresponding
+            # encoded values of cats in self.
+            cats = cats.astype(self.dtype)
+        except ValueError:
+            return Series(
+                utils.scalar_broadcast_to(
+                    na_sentinel, size=len(self), dtype=dtype
+                ),
+                index=self.index,
+                name=None,
+            )
+
         order = Series(cupy.arange(len(self)))
         codes = Series(cupy.arange(len(cats), dtype=dtype))
 
         value = DataFrame({"value": cats, "code": codes})
         codes = DataFrame(
-            {"value": self.copy(deep=False)._data.columns[0], "order": order}
+            {"value": self._data.columns[0].copy(deep=False), "order": order}
         )
         codes = codes.merge(value, on="value", how="left")
-        codes = codes.sort_values("order")["code"].fillna(na_sentinel)
 
-        cats.name = None  # because it was mutated above
+        # If the `code` column is full of nulls
+        # there is no need of sorting by `order`
+        if codes._data["code"].null_count == len(codes):
+            codes = codes["code"]
+        else:
+            codes = codes.sort_values("order")["code"]
```

Comment:
What's the motivation for adding this check?

Reply:
So this check is to avoid sorting by `order` when all of `code` column is `null`. The reason `code` column can all be null here is if we have non-intersecting categories in `self` & the categories passed(`cats`).
example:
```python
>>> s = cudf.Series([1,2,3])
>>> t = cudf.CategoricalDtype([5,6,7])
>>> s.astype(t)
> /conda/envs/cudf/lib/python3.7/site-packages/cudf/core/series.py(1685)label_encoding()
-> if codes._data["code"].null_count == len(codes):
(Pdb) codes
   value  order  code
0      1      0  null
1      2      1  null
2      3      2  null
```

Reply:
Is this a common operation that we actually need the short circuiting for? Otherwise this feels like a premature optimization to me.

Reply:
Had a discussion offline with @kkraus14 and decided to revert this change since the scenario is not a common one.